### ISSUE TITLE
SkyCoord: need to a documented way to handle proper motion correction, for case a coordinate has no proper motion

### ISSUE THREAD

**Issue by orionlee** [astropy/astropy#16541]
### What is the problem this feature will solve?

I have some logic that needs to apply proper motion correction, but the input coordinate occasionally has no proper motion.

In the current API, there is no documented way to handle this.

---

1. The input coordinate has been transformed to a `SkyCoord` object. When the coordinate has no proper motion:

```python
import astropy.units as u
from astropy.coordinates import SkyCoord
from astropy.time import Time

a_coord = SkyCoord(
    196.421 * u.deg, 18.01 * u.deg, frame="icrs",
)

new_coord = a_coord.apply_space_motion(new_obstime=Time(2000.0, format="jyear"))
```
It will raise `ValueError: SkyCoord requires velocity data to evolve the position.`

---

2. So I try to test if the coordinate as proper motion with `a_coord.pm_dec` or `a_coord.pm_ra_cosdec` attributes. But when the coordinate as no proper motion, accessing the attributes would raise `TypeError: Frame data has no associated differentials (i.e. the frame has no velocity data) - represent_as() only accepts a new representation.`

Similarly, accessing attribute .proper_motion` would raise `ValueError: Frame has no associated velocity (Differential) data information.`

---

3. For now, I end up testing if the coordinate has proper motion using an undocumented test, ` "s" in a_coord.frame.data.differentials ` , mimic undocumented internals: https://github.com/astropy/astropy/blob/37e4de53b1e06085363d655d81d0606bb1ea1cee/astropy/coordinates/baseframe.py#L1055

### Describe the desired outcome

It'd be great if
1. `apply_space_motion()` has an option of handling coordinate with no proper motion (essetinally a no-op), or
2. provide a documented way to test if a `SkyCoord` object has proper motion.

**Comment by neutrinoceros** [comment]
My prefered option would depend on the answer to the following question:
Is it always correct to assume that no velocity data means no proper motion ? In other words, would it be valid in any situation to have a `SkyCoord` object with *missing* velocity data, but who's proper motion cannot be assumed to be 0 ?

**Comment by orionlee** [comment]
Some cases to consider.

---
Case 1: If one applies proper motion to some catalog data (say, Gaia), where there are rows with missing proper motion, it's usually effectively 0 (due to astropy `Quantity` behavior). Example:
```python
import astropy.units as u
from astropy.coordinates import SkyCoord
from astropy.time import Time
from astropy.table import Table

#  simulate some catalog data with missing PM
tab = Table.read("""\
ra,dec,pm_ra,pm_dec
1.,45.,0.001,0.002
2.,4.4,,
""", format="ascii")
tab["ra"].unit = u.deg
tab["dec"].unit = u.deg
tab["pm_ra"].unit = u.milliarcsecond / u.year
tab["pm_dec"].unit = u.milliarcsecond / u.year

print(tab["pm_ra"])  # 2nd row is missing data
print(tab["pm_ra"].filled())  # the missing data is filled with some default large number
print(tab["pm_ra"].quantity)  # missing data is converted to 0

```
Ouput:
```pyhton
   pm_ra    
marcsec / yr
------------
       0.001
          --
   pm_ra    
marcsec / yr
------------
       0.001
       1e+20
[0.001 0.   ] marcsec / yr
```
Because missing PM data is converted to 0 in `Quantity`, effectively the `SkyCoord` object  constructed from the table would have an explicit 0 value for rows with missing PM.

```python
c = SkyCoord(tab["ra"].quantity, tab["dec"].quantity, 
             pm_ra_cosdec=tab["pm_ra"].quantity, pm_dec=tab["pm_dec"].quantity,
            frame='icrs', obstime=Time(2016, format="jyear"))
new_c = c.apply_space_motion(Time(2000, format="jyear"))
```

---

Case 2 (the context that led to this issue): Do proper motion correct on the coordinate of a target, specified in a TESS TargetPixelFile (postage stamp of TESS image data cutout for a specific target). The target's coordinate is specified in the FITS file header.  
For various reasons, proper motion would be missing in some data. In such case, a `SkyCoord` object without `pm_ra_cosdec` and `pm_dec` is created, and it leads to the problem.
(Explicitly set `pm_ra_cosdec` and `pm_dec` to 0 in creating `SkyCoord` objects would also solve by specific use case.)

**Comment by mhvk** [comment]
I think the current behaviour is correct in principle: if there is no proper motion data, that most likely is because it has not been measured, i.e., the proper motion is strictly unknown. I very much see the case, though, for having an option in `apply_space_motion` to state one wants to assume 0 if unknown.

For the test case, I think logically accessing `.proper_motion`, etc., should lead to an `AttributeError` so that `hasattr(sc, "proper_motion")` works (which just does a `try/except` internally -- obviously, that would be a reasonable work-around for now).

It is disturbing that right now the type of error depends on the attribute one tries:
```
hasattr(sc, "proper_motion")
# ValueError: Frame has no associated velocity (Differential) data information.
hasattr(sc, "pm_ra_cosdec")  # or "pm_dec" or "radial_velocity"
# TypeError: Frame data has no associated differentials (i.e. the frame has no velocity data) - represent_as() only accepts a new representation.
# hasattr(sc, "pm_ra")
False
```
The difference between `pm_ra` and `pm_ra_cosdec` is really rather bad; I have not given a proper motion, but somehow the system is nevertheless decided for be. I feel all of the above should give `False` -- which should be an easy change.

**Comment by orionlee** [comment]
@mvhk Is `Quantity` behavior of converting missing data (`np.ma.masked`) to 0 a deliberate behavior?

Treating missing proper motion as 0 is very convenient in the scenario of doing proper motion correction for catalog data (https://github.com/astropy/astropy/issues/16541#issuecomment-2152789148), I'm just wondering if it works this way because of some coincidence (and might accidentally change again in the future) or some deliberate choice.

**Comment by mhvk** [comment]
> @mvhk Is `Quantity`'s behavior of converting missing data (`np.ma.masked`) to 0 a deliberate behavior?

No, it is just that `np.ma.masked` is equivalent to `np.ma.MaskedArray(0.0, mask=True)` and that internally `Quantity` simply ignores the mask, doing the equivalent of 
```
In [5]: np.array(np.ma.masked)
Out[5]: array(0.)
```
In principle, now that `MaskedQuantity` exists, it may be more logical to return that - certainly less of a surprise than just stripping the mask!

**Comment by mhvk** [comment]
Actually, one we properly support masking in `SkyCoord`, at least the missing data problem will become a bit less bad: the result will be masked too (and things like `.filled` should become possible).

**Comment by orionlee** [comment]
Thanks mvhk for the clarification. 

For reference, I tried it with some actual Gaia DR3 data via Vizier, missing PM values are converted to `NaN` instead of `0`. 
So for the purpose of doing PM correction when data is available, one should use `tab["pm_ra"].filled(0).quantity` to explicitly treat missing data as 0 to be safe.


Example:
```python
from astropy.coordinates import SkyCoord
from astropy.time import Time
import astropy.units as u
from astroquery.vizier import Vizier

tab = Vizier().query_region(
    SkyCoord(324.55058921401 * u.deg, 41.68665352519 * u.deg, frame="icrs"),
    radius=6.5*u.arcsec,
    catalog="I/355/gaiadr3",
)[0]

# row 0 has missing PM
print(tab["Source", "RA_ICRS", "DE_ICRS", "pmRA", "pmDE"])

# missing PM is converted to NaN as quantity
print("pmRA: ", tab["pmRA"].quantity)

c = SkyCoord(tab["RA_ICRS"].quantity, tab["DE_ICRS"].quantity, 
             pm_ra_cosdec=tab["pmRA"].quantity, pm_dec=tab["pmDE"].quantity,
            frame='icrs', obstime=Time(2016, format="jyear"))

# the PM corrected corrected coordinate of row 0 is thus also nan
new_c = c.apply_space_motion(Time(2000, format="jyear"))
print(new_c)
```

Output
```python
       Source           RA_ICRS         DE_ICRS        pmRA      pmDE  
                          deg             deg        mas / yr  mas / yr
------------------- --------------- --------------- --------- ---------
1966708032133086848 324.55058921401  41.68665352519        --        --
1966708032138522240 324.55068474983  41.68843381032     3.013    -4.224

pmRA:  [  nan 3.013] mas / yr

<SkyCoord (ICRS): (ra, dec) in deg
    [(         nan,         nan), (324.55066682, 41.68845258)]
 (pm_ra_cosdec, pm_dec, radial_velocity) in (mas / yr, mas / yr, km / s)
    [(       nan,         nan,         nan),
     (3.01300088, -4.22399937, -0.00120716)]>
```
